vtkext/private/module/vtkF3DImguiConsole.cxx

```diff
@@ -0,0 +1,191 @@
+#include "vtkF3DImguiConsole.h"
+
+#include <vtkCommand.h>
+#include <vtkObjectFactory.h>
+
+#include <imgui.h>
+
+#include <array>
+
+struct vtkF3DImguiConsole::Internals
+{
+  enum class LogType
+  {
+    Log,
+    Warning,
+    Error,
+    Typed
+  };
+
+  std::vector<std::pair<LogType, std::string>> Logs;
+  std::array<char, 256> CurrentInput = {};
+  bool NewError = false;
+  bool NewWarning = false;
+};
+
+vtkStandardNewMacro(vtkF3DImguiConsole);
+
+//----------------------------------------------------------------------------
+vtkF3DImguiConsole::vtkF3DImguiConsole()
+  : Pimpl(new Internals())
+{
+}
+
+//----------------------------------------------------------------------------
+vtkF3DImguiConsole::~vtkF3DImguiConsole() = default;
+
+//----------------------------------------------------------------------------
+void vtkF3DImguiConsole::DisplayText(const char* text)
+{
+  switch (this->GetCurrentMessageType())
+  {
+    case vtkOutputWindow::MESSAGE_TYPE_ERROR:
+      this->Pimpl->Logs.emplace_back(std::make_pair(Internals::LogType::Error, text));
+      this->Pimpl->NewError = true;
+      break;
+    case vtkOutputWindow::MESSAGE_TYPE_WARNING:
+    case vtkOutputWindow::MESSAGE_TYPE_GENERIC_WARNING:
+      this->Pimpl->Logs.emplace_back(std::make_pair(Internals::LogType::Warning, text));
+      this->Pimpl->NewWarning = true;
+      break;
+    default:
+      this->Pimpl->Logs.emplace_back(std::make_pair(Internals::LogType::Log, text));
+  }
+
+  // still print text to std::cout in
+  // order to be able to test verbose outputs
+  this->Superclass::DisplayText(text);
+}
+
+//----------------------------------------------------------------------------
+void vtkF3DImguiConsole::ShowConsole()
+{
+  ImGuiViewport* viewport = ImGui::GetMainViewport();
+
+  constexpr float margin = 30.f;
+
+  this->Pimpl->NewError = false;
+  this->Pimpl->NewWarning = false;
+
+  ImGui::SetNextWindowPos(ImVec2(margin, margin));
+  ImGui::SetNextWindowSize(
+    ImVec2(viewport->WorkSize.x - 2.f * margin, viewport->WorkSize.y - 2.f * margin));
+  ImGui::SetNextWindowBgAlpha(0.9f);
+
+  ImGuiWindowFlags winFlags = ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoSavedSettings |
+    ImGuiWindowFlags_NoFocusOnAppearing | ImGuiWindowFlags_NoNav | ImGuiWindowFlags_NoMove;
+
+  // Since imgui has focus, it won't propagate the "Escape" key event to VTK
+  // So let's handle the console visibility here
+  if (ImGui::IsKeyPressed(ImGuiKey_Escape, false) && this->Pimpl->CurrentInput[0] == '\0')
+  {
+    this->InvokeEvent(vtkF3DImguiConsole::CommandEvent, "set ui.console false");
```

Comment:
@mwestphal is it fine? Or should we send a specific event like `vtkCommand::EnterEvent`/`vtkCommand::ExitEvent` and handle console visibility at `interactor_impl` level?

Reply:
definitely not what we want, this should not go through the command API.

Reply:
ok I will send specific events